src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking_FlowAnalysis.cs

```diff
@@ -1282,6 +1296,349 @@ static void G()
             comp.VerifyDiagnostics();
         }
 
+        [Fact]
+        public void IdentityConversion_NullCoalescingOperator_01()
+        {
+            var source =
+@"interface I<T> { }
+interface IIn<in T> { }
+interface IOut<out T> { }
+class C
+{
+    static void F(I<object>? x, I<object?> y)
+    {
+        I<object> z = x ?? y;
+        I<object?> w = y ?? x;
+    }
+    static void F(IIn<object>? x, IIn<object?> y)
+    {
+        IIn<object> z = x ?? y;
+        IIn<object?> w = y ?? x;
+    }
+    static void F(IOut<object>? x, IOut<object?> y)
+    {
+        IOut<object> z = x ?? y;
+        IOut<object?> w = y ?? x;
+    }
+}";
+            var comp = CreateStandardCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (8,28): warning CS8619: Nullability of reference types in value of type 'I<object?>' doesn't match target type 'I<object>'.
+                //         I<object> z = x ?? y;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "y").WithArguments("I<object?>", "I<object>").WithLocation(8, 28),
+                // (9,24): hidden CS8607: Expression is probably never null.
+                //         I<object?> w = y ?? x;
+                Diagnostic(ErrorCode.HDN_ExpressionIsProbablyNeverNull, "y").WithLocation(9, 24),
+                // (9,29): warning CS8619: Nullability of reference types in value of type 'I<object>' doesn't match target type 'I<object?>'.
+                //         I<object?> w = y ?? x;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "x").WithArguments("I<object>", "I<object?>").WithLocation(9, 29),
+                // (14,26): hidden CS8607: Expression is probably never null.
+                //         IIn<object?> w = y ?? x;
+                Diagnostic(ErrorCode.HDN_ExpressionIsProbablyNeverNull, "y").WithLocation(14, 26),
+                // (14,26): warning CS8619: Nullability of reference types in value of type 'IIn<object>' doesn't match target type 'IIn<object?>'.
+                //         IIn<object?> w = y ?? x;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "y ?? x").WithArguments("IIn<object>", "IIn<object?>").WithLocation(14, 26),
+                // (18,26): warning CS8619: Nullability of reference types in value of type 'IOut<object?>' doesn't match target type 'IOut<object>'.
+                //         IOut<object> z = x ?? y;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "x ?? y").WithArguments("IOut<object?>", "IOut<object>").WithLocation(18, 26),
+                // (19,27): hidden CS8607: Expression is probably never null.
+                //         IOut<object?> w = y ?? x;
+                Diagnostic(ErrorCode.HDN_ExpressionIsProbablyNeverNull, "y").WithLocation(19, 27));
+        }
+
+        [Fact]
+        public void IdentityConversion_NullCoalescingOperator_02()
```

Comment:
Comment tangential to this PR:
Currently, we're relying on consuming an expression to make its properties/state visible, and we're testing many permutations.
But the validation for this feature comes down to checking each kind of bound node for known incoming states and looking at what state and warnings comes out. 

What if we invested in an internal API to check the state of an expression? Could we reduce the test matrix that way?
We could come up with states worthy of testing and test them in method inference, separately from null-coalescing operator.

Reply:
Investigating checking inferred null state in the test helpers, based on annotations (comments) in the source.